### DISCUSSION TITLE
3D Obstacle Avoidance for Convex Shapes

### DISCUSSION THREAD

**Discussion by augustsb** [casadi/casadi#3717] [Q&A] [answered]
Hi,

I'm wondering if anyone knows any libraries that support casadi that can be used for 3D collision detection and obstacle avoidance? What I want to do is to represent the swept volume of my robot as a box and I would like to use convex polytopes as obstacles, but it seems this can't be done easily in casadi?
Alternatively, does anyone know any other optimization tools other than casadi that will be better for this?

**Comment by charleskhazoom** [comment, marked as answer]
There is  The Flexible Collision Library:
https://github.com/flexible-collision-library/fcl

There is also an extension to it which seems to be improved and still maintained : 

https://github.com/humanoid-path-planner/hpp-fcl 

For obstacle avoidance, depending on your chosen method, you may need the gradients of the signed-distance functions. This is often not readily differentiable except for spheres. Thus, it does not necessarily work well with Casadi. But approximations such as the ones used here work well in practice : https://arxiv.org/abs/2207.00692

**Reply by augustsb** [reply]
Ok, but is it possible to use FCL to compute the distances as external function and then somehow approximate the derivatives using finite differences or something?

**Reply by charleskhazoom** [reply]
FCL can give you the distances between two shapes as well as the closest points on those shapes. Once you have the closest points, the approximations for computing J_AB in https://arxiv.org/abs/2207.00692 can be applied. To compute J_A, you could use Casadi's jacobian function to differentiate the closest point value on the object with respect to the pose of the object.

**Comment by jaydguthrie** [comment]
Late to this discussion, but wanted to point out a simpler method that is fully differentiable. Full disclosure, it's my own work. :) 

If you want to add collision avoidance constraints (or more generally, signed distance constraints) to a trajectory optimization / motion planning problem, you can do this easily with CasADi for arbitrary convex shapes. There is no need to use an external library (e.g. FCL) for computing the signed distance function and then attempting to differentiate this function. Rather you introduce an additional vector variable at each time step of your problem along with constraints that make the variable define a separating hyperplane between the robot and the obstacle. You can specify the amount of separation required (e.g. the minimum signed distance). Additionally, you can represent swept volume in this framework to ensure continuous collision avoidance (not just collision avoidance at discrete time steps).

The formulation is quite simple.
- For polytopes and ellipsoids, see Section 4. Examples 1 & 2 of [1]
- For more general convex sets described by second-order cone constraints, polynomial inequalities, Minkowski sums, etc. - see Ch. 5 of [2]

[1] J. Guthrie, "A Differentiable Signed Distance Representation for Continuous Collision Avoidance in Optimization-Based Motion Planning," 2022 IEEE 61st Conference on Decision and Control (CDC)
[PDF](https://arxiv.org/pdf/2302.09704)

[2] J. Guthrie, “Novel representations of semialgebraic sets arising in planning and control,” Ph.D. dissertation
[PDF](https://jscholarship.library.jhu.edu/bitstreams/7a09ce11-2a23-4772-8c66-350477597f17/download)

**Comment by peterminh227** [comment]
Hi Jay, 
Thanks for your answer. I have checked two of your works; they are great! 
I wonder if you have a source code for a simple example? 
I have checked the CCA repo in your first link. However, it is empty.
